### PR TITLE
feat: difficulty adjustment rfc version

### PR DESCRIPTION
**Pull request by zhangsoledad**
Apply new difficulty adjustment mechanism
https://github.com/nervosnetwork/rfcs/blob/master/rfcs/0020-ckb-consensus-protocol/0020-ckb-consensus-protocol.md#dynamic-difficulty-adjustment-mechanism

### CODE REVIEW

File: spec/src/consensus.rs [hunk 1]
```diff
@@ -340,82 +377,105 @@ impl Consensus {
...
-                last_epoch.number() + 1,    // number
-                block_reward,
-                remainder_reward,           // remainder_reward
-                header.hash().to_owned(),   // last_block_hash_in_previous_epoch
-                header.number() + 1,        // start
-                next_epoch_length,          // length
-                difficulty                  // difficulty,
-            )
+            let denominator = (target_recip + 1) * next_epoch_length / target_recip;
+            diff_numerator / denominator
```
Comment (janx) [inline spec/src/consensus.rs]: The calculation seems problematic here.

File: spec/src/consensus.rs [hunk 2]
```diff
@@ -340,82 +377,105 @@ impl Consensus {
...
-                header.number() + 1,         // start
-                next_epoch_length,           // length
-                difficulty                   // difficulty,
-            )
+            self.bounding_epoch_length(raw_next_epoch_length)
+        };
+
+        // (3) Determining the Next Epoch’s Difficulty
+        let diff_numerator = &adjusted_last_epoch_hash_rate * epoch_duration_target;
+        let next_epoch_diff = if bound {
```
Comment (janx) [inline spec/src/consensus.rs]: Using the single unified equation is better than calculate differently on different cases.

File: spec/src/consensus.rs [hunk 3]
```diff
@@ -23,17 +23,21 @@ pub(crate) const MEDIAN_TIME_BLOCK_COUNT: usize = 11;
...
+pub(crate) const EPOCH_DURATION_TARGET: u64 = 4 * 60 * 60; // 4 hours, unit: second
+pub(crate) const MILLISECOND_TO_SECOND: u64 = 1000;
 pub(crate) const MAX_EPOCH_LENGTH: u64 = EPOCH_DURATION_TARGET / MIN_BLOCK_INTERVAL; // 1800
 pub(crate) const MIN_EPOCH_LENGTH: u64 = EPOCH_DURATION_TARGET / MAX_BLOCK_INTERVAL; // 480
 
-// We choose 1250 because it is largest number between MIN_EPOCH_LENGTH and MAX_EPOCH_LENGTH that
-// can divide DEFAULT_EPOCH_REWARD.
-pub(crate) const GENESIS_EPOCH_LENGTH: u64 = 1_250;
+// We choose 1_000 because it is largest number between MIN_EPOCH_LENGTH and MAX_EPOCH_LENGTH that
+// can divide DEFAULT_EPOCH_REWARD and ORPHAN_RATE_TARGET_RECIP.
```
[the remark this answers is not in the document]
Comment (zhangsoledad) [reply spec/src/consensus.rs]: you mixed up `divide` with `divide by `.

**Comment by u2** [comment]
Looks good for me, except the `Determining the Next Epoch’s Difficulty` part which I did not understand.

**Comment by janx** [comment]
@u2 please don't approve unless you have no further problem about the code, otherwise code review is meaningless.

File: spec/src/consensus.rs [hunk 4]
```diff
@@ -23,17 +23,21 @@ pub(crate) const MEDIAN_TIME_BLOCK_COUNT: usize = 11;
...
+pub(crate) const TAU: u64 = 2;
 
-const MAX_BLOCK_INTERVAL: u64 = 30 * 1000; // 30s
-const MIN_BLOCK_INTERVAL: u64 = 8 * 1000; // 8s
+const MAX_BLOCK_INTERVAL: u64 = 30; // 30s
+const MIN_BLOCK_INTERVAL: u64 = 8; // 8s
 const TWO_IN_TWO_OUT_CYCLES: Cycle = 2_580_000;
-pub(crate) const EPOCH_DURATION_TARGET: u64 = 4 * 60 * 60 * 1000; // 4 hours
+pub(crate) const EPOCH_DURATION_TARGET: u64 = 4 * 60 * 60; // 4 hours, unit: second
+pub(crate) const MILLISECOND_TO_SECOND: u64 = 1000;
```
Comment (doitian) [inline spec/src/consensus.rs]: You get 0.001 second when converting a millisecond to second. 

```suggestion
pub(crate) const MILLISECONDS_IN_A_SECOND: u64 = 1000;
```

File: util/rational/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,548 @@
...
+        self.denom = &self.denom / &g;
+    }
+}
+
+// a/b * c/d = (a*c)/(b*d)
+impl Mul<&RationalU256> for &RationalU256 {
+    type Output = RationalU256;
+    #[inline]
+    fn mul(self, rhs: &RationalU256) -> RationalU256 {
+        RationalU256::new(&self.numer * &rhs.numer, &self.denom * &rhs.denom)
```
Comment (doitian) [inline util/rational/src/lib.rs]: What's the behavior when the U256 multiplication overflows?

Comment (zhangsoledad) [reply util/rational/src/lib.rs]: panic whether release or debug

File: spec/src/consensus.rs [hunk 6]
```diff
@@ -340,82 +383,124 @@ impl Consensus {
...
+                (diff_numerator / denominator).into_u256()
+            } else {
+                let orphan_rate_estimation = RationalU256::one()
+                    / (((&last_orphan_rate + U256::one())
+                        * &epoch_duration_target_u256
+                        * &last_epoch_length_u256
+                        / (&last_orphan_rate * &last_epoch_duration * &next_epoch_length_u256))
+                        .saturating_sub_u256(U256::one()));
+
+                let denominator = if orphan_rate_estimation.is_zero() {
```
Comment (janx) [inline spec/src/consensus.rs]: Is it possible `orphan_rate_estimation` is negative or larger than `orphan_rate_target`?

Comment (zhangsoledad) [reply spec/src/consensus.rs]: @nirenzang is not sure about this, so we use `saturating_sub` here.

Comment (janx) [reply spec/src/consensus.rs]: What if the `orphan_rate_estimation` is larger than `orphan_rate_target`?

Comment (zhangsoledad) [reply spec/src/consensus.rs]: It is possible `orphan_rate_estimation` than `orphan_rate_target`, is there a problem?

File: spec/src/consensus.rs [hunk 7]
```diff
@@ -21,19 +22,24 @@ pub(crate) const CELLBASE_MATURITY: BlockNumber = 100;
...
+pub(crate) const EPOCH_DURATION_TARGET: u64 = 4 * 60 * 60; // 4 hours, unit: second
+pub(crate) const MILLISECONDS_IN_A_SECOND: u64 = 1000;
 pub(crate) const MAX_EPOCH_LENGTH: u64 = EPOCH_DURATION_TARGET / MIN_BLOCK_INTERVAL; // 1800
 pub(crate) const MIN_EPOCH_LENGTH: u64 = EPOCH_DURATION_TARGET / MAX_BLOCK_INTERVAL; // 480
 
-// We choose 1250 because it is largest number between MIN_EPOCH_LENGTH and MAX_EPOCH_LENGTH that
-// can divide DEFAULT_EPOCH_REWARD.
-pub(crate) const GENESIS_EPOCH_LENGTH: u64 = 1_250;
+// We choose 1_000 because it is largest number between MIN_EPOCH_LENGTH and MAX_EPOCH_LENGTH that
+// can divide DEFAULT_EPOCH_REWARD and can divide by ORPHAN_RATE_TARGET_RECIP.
```
Comment (quake) [inline spec/src/consensus.rs]:
```suggestion
// can divide DEFAULT_EPOCH_REWARD and can be divided by ORPHAN_RATE_TARGET_RECIP.
```

File: spec/src/consensus.rs (hunk 6 above)
Comment (janx) [reply spec/src/consensus.rs]: I can't find a problem here.

**Pull request merged**